### PR TITLE
Fix memory leaks in pubsub reporter.

### PR DESCRIPTION
**Pull request by cjwagner**
I think this is the root cause of the memory leak we've observed in Crier. I looked through the controller and all the reporters, this is the only memory leak I can spot, but I think it explains why we've only had these leaks grow out of control on an instance where pubsub is more heavily used.

/assign @fejta @chaodaiG 
/kind bug

### CODE REVIEW

File: prow/crier/reporters/pubsub/reporter.go [hunk 1]
```diff
@@ -98,12 +98,16 @@ func (c *Client) Report(_ *logrus.Entry, pj *prowapi.ProwJob) ([]*prowapi.ProwJo
 	message := c.generateMessageFromPJ(pj)
 
 	ctx := context.Background()
+	// TODO: Consider caching the pubsub client.
 	client, err := pubsub.NewClient(ctx, message.Project)
```
Comment (alvaroaleman) [inline prow/crier/reporters/pubsub/reporter.go:102]: The client also has a `Close()`, mostly to shutdown its connection pool apparently, maybe `defer` that as well?

Comment (cjwagner) [reply prow/crier/reporters/pubsub/reporter.go:102]: Yes, I'm already (sneakily) doing that on line 107 within the log statement. 😄

Comment (alvaroaleman) [reply prow/crier/reporters/pubsub/reporter.go:102]: Ah, missed that. I am pretty sure this doesn't do what you think though, the log statement itself does get deferred, but the resolving of its argument does not, hence you will close the client in line 107 and not via the `defer`

File: prow/crier/reporters/pubsub/reporter.go [hunk 2]
```diff
@@ -98,12 +98,16 @@ func (c *Client) Report(_ *logrus.Entry, pj *prowapi.ProwJob) ([]*prowapi.ProwJo
...
 
 	if err != nil {
 		return nil, nil, fmt.Errorf("could not create pubsub Client: %v", err)
 	}
+	defer logrus.WithError(client.Close()).Debug("Closed pubsub client.")
```
Comment (fejta) [inline prow/crier/reporters/pubsub/reporter.go:107]:
```golang
defer func() {
  logrus.WithError().Debug()
}()
```

should address Alvaro's point

File: prow/crier/reporters/pubsub/reporter.go [hunk 3]
```diff
@@ -98,12 +98,16 @@ func (c *Client) Report(_ *logrus.Entry, pj *prowapi.ProwJob) ([]*prowapi.ProwJo
 	message := c.generateMessageFromPJ(pj)
 
 	ctx := context.Background()
+	// TODO: Consider caching the pubsub client.
```
Comment (fejta) [inline prow/crier/reporters/pubsub/reporter.go:101]: This definitely seems like an excellent idea for a subsequent PR. Current design is incredibly wasteful.

File: prow/crier/reporters/pubsub/reporter.go (hunk 1 above)
Comment (cjwagner) [reply prow/crier/reporters/pubsub/reporter.go:102]: Good catch, I thought the arguments were evaluated lazily. Fixed.

**Comment by cjwagner** [comment]
PTAL
/cc @chases2

**Review by chases2 — APPROVED** [review]

**Pull request merged**